tasks/evolutionary-algorithm/src/main.rs

```diff
@@ -4,94 +4,95 @@
 
-    println!("{}", target);
-    println!("{}", parent);
-    while fitness(&target, &parent) != 0 {
-        let mut sentences: HashMap<u32, String> = HashMap::new();
-        let mut f_min: u32 = 30;
+    // Generate first sentence, mutating each character
+    let start = mutate(&mut rng, target, 1); // 1/1 = 1 = 100%
 
-        counter += 1;
+    println!("{}", target);
+    println!("{}", start);
 
-        for _ in 0..nb_copy {
-            let sentence = mutate(&parent, mutation_rate);
-            let f = fitness(&target, &sentence);
+    evolve(&mut rng, target, start, copies, mutation_rate);
+}
 
-            sentences.insert(f, sentence);
+/// Evolution algorithm
+///
+/// Evolves `parent` to match `target`.  Returns the number of evolutions performed.
+fn evolve<R: Rng>(
+    rng: &mut R,
+    target: &str,
+    mut parent: String,
+    copies: usize,
+    mutation_rate: u32,
+) -> usize {
+    let mut counter = 0;
+    let mut parent_fitness = target.len() + 1;
 
-            if f < f_min {
-                f_min = f
-            }
-        }
+    loop {
+        counter += 1;
 
-        if fitness(&target, &parent) > f_min {
-            match sentences.get(&f_min) {
-                Some(s) => {
-                    parent = s.clone();
-                    println!("{} : {}", parent, counter);
-                }
-                None => panic!("Error, fitness minimum but no sentence."),
+        let (best_fitness, best_sentence) = (0..copies)
+            .map(|_| {
+                // Copy and mutate a new sentence.
+                let sentence = mutate(rng, &parent, mutation_rate);
+                // Find the fitness of the new mutation
+                (fitness(target, &sentence), sentence)
+            })
+            .min_by_key(|&(f, _)| f) // find the closest mutation to the target
+            .unwrap(); // fails if `copies == 0`
+
+        // If the best mutation of this generation is better than `parent` then "the fittest
+        // survives" and the next parent becomes the best of this generation.
+        if best_fitness < parent_fitness {
+            parent = best_sentence;
+            parent_fitness = best_fitness;
+            println!(
+                "{} : generation {} with fitness {}",
+                parent, counter, best_fitness
+            );
+
+            if best_fitness == 0 {
+                return counter;
             }
         }
     }
 }
 
-/// Computes the fitness of a sentence against a target string.
-fn fitness(target: &str, sentence: &str) -> u32 {
-    target.chars()
-        .zip(sentence.chars())
-        .fold(0, |acc, (c1, c2)| {
-            if c1 != c2 {
-                acc + 1
-            } else {
-                acc
-            }
-        })
+/// Computes the fitness of a sentence against a target string, returning the number of
+/// incorrect characters.
+fn fitness(target: &str, sentence: &str) -> usize {
+    let zip = sentence.chars().zip(target.chars());
```

Comment:
Why not keep this as one chain?

Reply:
I'm not sure what you mean.  Do you mean keeping the `zip` in `evolve`?

Reply:
Can it be written like so:

```
sentence.chars()
    .zip(target.chars())
    .filter(|&(c1, c2)| c1 != c2)
    .count()
````

Reply:
Oh yeah more idiomatic syntax is probably better.  Thanks.